src/selfplay/loop.cc

```diff
@@ -495,72 +502,115 @@ void ProcessFile(const std::string& file, SyzygyTablebase* tablebase,
-      // for 3 piece no-pawn positions only.
-      PopulateBoard(pblczero::NetworkFormat::INPUT_CLASSICAL_112_PLANE,
-                    PlanesFromTrainingData(fileContents[0]), &board, &rule50ply,
-                    &gameply);
-      history.Reset(board, rule50ply, gameply);
-      for (int i = 0; i < moves.size(); i++) {
-        history.Append(moves[i]);
-        const auto& board = history.Last().GetBoard();
-        if (board.castlings().no_legal_castle() &&
-            (board.ours() | board.theirs()).count() <= 3 &&
-            board.pawns().empty()) {
-          ProbeState state;
-          WDLScore wdl = tablebase->probe_wdl(history.Last(), &state);
-          // Only fail state means the WDL is wrong, probe_wdl may produce
-          // correct result with a stat other than OK.
-          if (state != FAIL) {
-            int8_t score_to_apply = 0;
-            if (wdl == WDL_WIN) {
-              score_to_apply = 1;
-            } else if (wdl == WDL_LOSS) {
-              score_to_apply = -1;
-            }
-            // No point updating for draws.
-            if (score_to_apply == 0) continue;
-            // Any repetitions in the history since last 50 ply makes it risky
-            // to assume dtz is still correct.
-            int steps = history.Last().GetNoCaptureNoPawnPly();
-            bool no_reps = true;
-            for (int i = 0; i < steps; i++) {
-              // If game started from non-zero 50 move rule, this could
-              // underflow. Only safe option is to assume there were repetitions
-              // before this point.
-              if (history.GetLength() - i - 1 < 0) {
-                no_reps = false;
-                break;
-              }
-              if (history.GetPositionAt(history.GetLength() - i - 1)
-                      .GetRepetitions() != 0) {
-                no_reps = false;
-                break;
+
+      // Correct plies_left using Gaviota TBs for 5 piece and less positions.
+      if (gaviotaEnabled && !all_draws) {
+        PopulateBoard(pblczero::NetworkFormat::INPUT_CLASSICAL_112_PLANE,
+                      PlanesFromTrainingData(fileContents[0]), &board,
+                      &rule50ply, &gameply);
+        history.Reset(board, rule50ply, gameply);
+        int last_rescore = 0;
+        for (int i = 0; i < moves.size(); i++) {
+          history.Append(moves[i]);
+          const auto& board = history.Last().GetBoard();
+
+          // Gaviota TBs don't have 50 move rule.
+          // Only consider positions that are not draw after rescoring.
+          if ((fileContents[i + 1].result != 0) &&
+              board.castlings().no_legal_castle() &&
+            (board.ours() | board.theirs()).count() <= 5) {
+
+            std::vector<int> dtms;
+            unsigned int info;
+            unsigned int dtm;
+            gaviota_tb_probe_hard(history.Last(), info, dtm);
+            if (info == tb_WMATE || info == tb_BMATE) {
```

Comment:
I think there should be more conditionals here.
If the DTM is over 100 ply, then its likely going to be lower than the DTM50 - so it can only serve as a lower bound, not an exact training target and we should only increase the move_count if it is lower than the DTM, rather than simply perform equality.
Additionally not checking for repeats puts us at risk here.  A position can not be rescored as a draw, but still potentially be one if a repeat has taken place if the WDL/DTZ thinks its a win if you ignore repeats (since we don't know if its a win or a draw, so we only rescore if its incorrectly marked as a loss, so if the actual game play was a win, the win will stand as the training target).  If there is a repeat we both don't know if its actually even possible to win anymore, and whether the DTM presented is plausible, even if it is under 100.  I think similar logic as from below needs to be brought up here.  Any repeats means no move count adjusting until the next zeroing move, then the intermediate moves can be adjusted based on that point.